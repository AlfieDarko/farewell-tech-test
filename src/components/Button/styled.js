import styled from "styled-components"
import { motion } from "framer-motion"
import { COLOR, BORDER, FONT, SPACING } from "../../constants"

export const Button = styled(motion.button)`
  padding: ${SPACING.XS} ${SPACING.M};
  cursor: pointer;
  outline: 0;
  border: 0;
  background-color: ${COLOR.ACCENT.PRIMARY};
  border-radius: ${BORDER.RADIUS.S};
  color: ${COLOR.BLACK};
  font-family: ${FONT.FAMILY.BODY};
  font-size: ${FONT.SIZE.M};
  text-decoration: none;
  min-width: 8em;
  max-height: 2em;

  &:hover {
    text-decoration: none;
  }

  &:focus {
    outline: 1px dashed ${COLOR.ACCENT.SECONDARY};
  }

  &:active {
    outline: 1px dashed ${COLOR.ACCENT.PRIMARY};
  }

  &:disabled {
    pointer-events: none;
    user-select: none;
  }
`

export const HomerSpinner = styled(motion.img)`
  height: 4em;
  position: relative;
  top: -${SPACING.M};
`
