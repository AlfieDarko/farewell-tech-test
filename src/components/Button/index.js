import React from "react"
import HomerSVG from "./assets/homer-svg.svg"
import * as Styled from "./styled"

const Button = (props) => {
  const { isLoading, children } = props

  return (
    <Styled.Button
      whileTap={{ scale: 0.9 }}
      transition={{ ease: "easeOut" }}
      disabled={isLoading}
      {...props}
    >
      {isLoading && (
        <Styled.HomerSpinner
          role="progressbar"
          aria-valuetext="Loading"
          transition={{ ease: "easeOut", repeat: "Infinity", duration: 2 }}
          src={HomerSVG}
          animate={{
            rotate: [0, 360],
            scale: [0.8, 0.5, 0.8],
          }}
        />
      )}
      {!isLoading && children}
    </Styled.Button>
  )
}

export default Button
