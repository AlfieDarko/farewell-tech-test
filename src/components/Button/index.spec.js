import React from "react"
import { render, screen } from "@testing-library/react"
import "@testing-library/jest-dom/extend-expect"
import Button from "./index"

describe("Button Component", () => {
  it("Should find Button Component by its data attribute", () => {
    render(<Button data-testid={"button"}>"Hello"</Button>)
    expect(screen.getByTestId("button")).toBeInTheDocument()
  })

  it("Should find spinner when in loading state ", async () => {
    render(
      <Button data-testid="button" isLoading={true}>
        "Hello"
      </Button>
    )
    expect(screen.getByRole("progressbar"))
  })

  it("Should match the snapshot", () => {
    const { asFragment } = render(
      <Button data-testid={"button"}>"Hello"</Button>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
