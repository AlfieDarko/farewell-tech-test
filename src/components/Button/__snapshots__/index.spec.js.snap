// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component Should match the snapshot 1`] = `
<DocumentFragment>
  <button
    class="sc-bdnylx bKxXSk"
    data-testid="button"
  >
    "Hello"
  </button>
</DocumentFragment>
`;
